// amplify-category-function/lib/provider-utils/awscloudformation/utils/cloudformationHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.constructCloudWatchEventComponent = exports.constructCFModelTableNameComponent = exports.constructCFModelTableArnComponent = exports.getIAMPolicies = exports.getNewCFNParameters = exports.getNewCFNEnvVariables = void 0;
const constants_1 = require("./constants");
const appSyncHelper_1 = require("./appSyncHelper");
function getNewCFNEnvVariables(oldCFNEnvVariables, currentDefaults, newCFNEnvVariables, newDefaults, apiResourceName) {
    const currentResources = [];
    const newResources = [];
    let deletedResources = [];
    const categorySet = new Set();
    if (currentDefaults.permissions) {
        Object.keys(currentDefaults.permissions).forEach(category => {
            Object.keys(currentDefaults.permissions[category]).forEach(resourceName => {
                categorySet.add(category);
                currentResources.push(`${category.toUpperCase()}_${resourceName.toUpperCase()}_`);
            });
        });
    }
    if (newDefaults.permissions) {
        Object.keys(newDefaults.permissions).forEach(category => {
            Object.keys(newDefaults.permissions[category]).forEach(resourceName => {
                newResources.push(`${category.toUpperCase()}_${resourceName.toUpperCase()}_`);
            });
        });
    }
    if (apiResourceName) {
        apiResourceAddCheck(currentResources, newResources, apiResourceName, categorySet, true);
    }
    currentResources.forEach(resourceName => {
        if (newResources.indexOf(resourceName) === -1) {
            deletedResources.push(resourceName);
        }
    });
    const deleteAppSyncTableResources = deletedResources.filter(resource => resource.includes(constants_1.appsyncTableSuffix.toUpperCase()));
    deletedResources = deletedResources.filter(resource => !resource.includes(constants_1.appsyncTableSuffix.toUpperCase()));
    deleteAppSyncTableResources.forEach(table => {
        const appsyncResourceName = appSyncHelper_1.getAppSyncResourceName();
        const replacementTableSuffix = `:${constants_1.appsyncTableSuffix.toUpperCase()}_`;
        const modelEnvPrefix = `API_${appsyncResourceName.toUpperCase()}_${table
            .replace(replacementTableSuffix, 'TABLE')
            .replace('STORAGE_', '')}`;
        const modelEnvNameKey = `${modelEnvPrefix}_NAME`;
        const modelEnvArnKey = `${modelEnvPrefix}_ARN`;
        deletedResources.push(modelEnvNameKey);
        deletedResources.push(modelEnvArnKey);
    });
    const toBeDeletedEnvVariables = [];
    Object.keys(oldCFNEnvVariables).forEach(envVar => {
        for (let i = 0; i < deletedResources.length; i += 1) {
            if (envVar.includes(deletedResources[i])) {
                toBeDeletedEnvVariables.push(envVar);
                break;
            }
        }
    });
    toBeDeletedEnvVariables.forEach(envVar => {
        delete oldCFNEnvVariables[envVar];
    });
    Object.assign(oldCFNEnvVariables, newCFNEnvVariables);
    return oldCFNEnvVariables;
}
exports.getNewCFNEnvVariables = getNewCFNEnvVariables;
function getNewCFNParameters(oldCFNParameters, currentDefaults, newCFNResourceParameters, newDefaults, apiResourceName) {
    const currentResources = [];
    const newResources = [];
    const deletedResources = [];
    const categorySet = new Set();
    if (currentDefaults.permissions) {
        Object.keys(currentDefaults.permissions).forEach(category => {
            Object.keys(currentDefaults.permissions[category]).forEach(resourceName => {
                categorySet.add(category);
                currentResources.push(`${category}${resourceName}`);
            });
        });
    }
    if (newDefaults.permissions) {
        Object.keys(newDefaults.permissions).forEach(category => {
            Object.keys(newDefaults.permissions[category]).forEach(resourceName => {
                newResources.push(`${category}${resourceName}`);
            });
        });
    }
    if (apiResourceName) {
        apiResourceAddCheck(currentResources, newResources, apiResourceName, categorySet, false);
    }
    currentResources.forEach(resourceName => {
        if (newResources.indexOf(resourceName) === -1) {
            deletedResources.push(resourceName);
        }
    });
    const toBeDeletedParameters = [];
    Object.keys(oldCFNParameters).forEach(parameter => {
        for (let i = 0; i < deletedResources.length; i += 1) {
            if (parameter.includes(deletedResources[i])) {
                toBeDeletedParameters.push(parameter);
                break;
            }
        }
    });
    toBeDeletedParameters.forEach(parameter => {
        delete oldCFNParameters[parameter];
    });
    Object.assign(oldCFNParameters, newCFNResourceParameters);
    return oldCFNParameters;
}
exports.getNewCFNParameters = getNewCFNParameters;
function getIAMPolicies(resourceName, crudOptions) {
    let policy = {};
    const actions = [];
    crudOptions.forEach(crudOption => {
        switch (crudOption) {
            case 'create':
                actions.push('lambda:Create*', 'lambda:Put*', 'lambda:Add*');
                break;
            case 'update':
                actions.push('lambda:Update*');
                break;
            case 'read':
                actions.push('lambda:Get*', 'lambda:List*', 'lambda:Invoke*');
                break;
            case 'delete':
                actions.push('lambda:Delete*', 'lambda:Remove*');
                break;
            default:
                console.log(`${crudOption} not supported`);
        }
    });
    policy = {
        Effect: 'Allow',
        Action: actions,
        Resource: [
            {
                'Fn::Join': [
                    '',
                    [
                        'arn:aws:lambda:',
                        {
                            Ref: 'AWS::Region',
                        },
                        ':',
                        { Ref: 'AWS::AccountId' },
                        ':function:',
                        {
                            Ref: `${constants_1.categoryName}${resourceName}Name`,
                        },
                    ],
                ],
            },
        ],
    };
    const attributes = ['Name'];
    return { policy, attributes };
}
exports.getIAMPolicies = getIAMPolicies;
function constructCFModelTableArnComponent(appsyncResourceName, resourceName, appsyncTableSuffix) {
    return [
        'arn:aws:dynamodb:',
        { Ref: 'AWS::Region' },
        ':',
        { Ref: 'AWS::AccountId' },
        ':table/',
        constructCFModelTableNameComponent(appsyncResourceName, resourceName, appsyncTableSuffix),
    ];
}
exports.constructCFModelTableArnComponent = constructCFModelTableArnComponent;
function constructCFModelTableNameComponent(appsyncResourceName, resourceName, appsyncTableSuffix) {
    return {
        'Fn::ImportValue': {
            'Fn::Sub': `\${api${appsyncResourceName}GraphQLAPIIdOutput}:GetAtt:${resourceName.replace(`:${appsyncTableSuffix}`, 'Table')}:Name`,
        },
    };
}
exports.constructCFModelTableNameComponent = constructCFModelTableNameComponent;
function constructCloudWatchEventComponent(cfnFilePath, cfnContent) {
    cfnContent.Resources.CloudWatchEvent = {
        Type: 'AWS::Events::Rule',
        Properties: {
            Description: 'Schedule rule for Lambda',
            ScheduleExpression: {
                Ref: 'CloudWatchRule',
            },
            State: 'ENABLED',
            Targets: [
                {
                    Arn: { 'Fn::GetAtt': ['LambdaFunction', 'Arn'] },
                    Id: {
                        Ref: 'LambdaFunction',
                    },
                },
            ],
        },
    };
    cfnContent.Resources.PermissionForEventsToInvokeLambda = {
        Type: 'AWS::Lambda::Permission',
        Properties: {
            FunctionName: {
                Ref: 'LambdaFunction',
            },
            Action: 'lambda:InvokeFunction',
            Principal: 'events.amazonaws.com',
            SourceArn: { 'Fn::GetAtt': ['CloudWatchEvent', 'Arn'] },
        },
    };
    cfnContent.Outputs.CloudWatchEventRule = {
        Value: {
            Ref: 'CloudWatchEvent',
        },
    };
    if (cfnContent.Parameters.CloudWatchRule === undefined) {
        cfnContent.Parameters.CloudWatchRule = {
            Type: 'String',
            Default: 'NONE',
            Description: ' Schedule Expression',
        };
    }
}
exports.constructCloudWatchEventComponent = constructCloudWatchEventComponent;
function apiResourceAddCheck(currentResources, newResources, apiResourceName, resourceSet, isEnvParams) {
    const apiAddFlag = resourceSet.has('api') || !newResources.find(resource => resource.includes('storage'));
    if (apiAddFlag) {
        isEnvParams ? currentResources.push(`API_${apiResourceName.toUpperCase()}_`) : currentResources.push(`api${apiResourceName}`);
    }
}
