// amplify-category-function/lib/provider-utils/awscloudformation/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.advancedSettingsList = exports.CronExpressionsMode = exports.ServiceName = exports.lambdaLayerSetting = exports.cronJobSetting = exports.resourceAccessSetting = exports.appsyncTableSuffix = exports.provider = exports.parametersFileName = exports.layerParametersFileName = exports.functionParametersFileName = exports.chooseServiceMessageUpdate = exports.chooseServiceMessageAdd = exports.categoryName = void 0;
const constants_1 = require("../../../constants");
exports.categoryName = constants_1.category;
exports.chooseServiceMessageAdd = 'Select which capability you want to add:';
exports.chooseServiceMessageUpdate = 'Select which capability you want to update:';
exports.functionParametersFileName = 'function-parameters.json';
exports.layerParametersFileName = 'layer-parameters.json';
exports.parametersFileName = 'parameters.json';
exports.provider = 'awscloudformation';
exports.appsyncTableSuffix = '@model(appsync)';
exports.resourceAccessSetting = 'Resource access permissions';
exports.cronJobSetting = 'Scheduled recurring invocation';
exports.lambdaLayerSetting = 'Lambda layers configuration';
var ServiceName;
(function (ServiceName) {
    ServiceName["LambdaFunction"] = "Lambda";
    ServiceName["LambdaLayer"] = "LambdaLayer";
})(ServiceName = exports.ServiceName || (exports.ServiceName = {}));
var CronExpressionsMode;
(function (CronExpressionsMode) {
    CronExpressionsMode["Minutes"] = "Minutes";
    CronExpressionsMode["Hourly"] = "Hourly";
    CronExpressionsMode["Daily"] = "Daily";
    CronExpressionsMode["Weekly"] = "Weekly";
    CronExpressionsMode["Monthly"] = "Monthly";
    CronExpressionsMode["Yearly"] = "Yearly";
    CronExpressionsMode["Custom"] = "Custom AWS cron expression";
})(CronExpressionsMode = exports.CronExpressionsMode || (exports.CronExpressionsMode = {}));
exports.advancedSettingsList = [exports.resourceAccessSetting, exports.cronJobSetting, exports.lambdaLayerSetting];
