// amplify-category-function/lib/commands/function/remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../../constants");
const storeResources_1 = require("../../provider-utils/awscloudformation/utils/storeResources");
const subcommand = 'remove';
module.exports = {
    name: subcommand,
    run: async (context) => {
        const { amplify, parameters } = context;
        const resourceName = parameters.first;
        return amplify
            .removeResource(context, constants_1.category, resourceName, {
            serviceDeletionInfo: {
                LambdaLayer: 'When you delete a layer version, you can no longer configure functions to use it.\nHowever, any function that already uses the layer version continues to have access to it.',
            },
            serviceSuffix: { Lambda: '(function)', LambdaLayer: '(layer)' },
        })
            .then((resource) => {
            if (resource.service === "LambdaLayer") {
                storeResources_1.removeLayerArtifacts(context, resource.resourceName);
            }
        })
            .catch(err => {
            context.print.info(err.stack);
            context.print.error('An error occurred when removing the function resource');
            context.usageData.emitError(err);
            process.exitCode = 1;
        });
    },
};
