// amplify-nodejs-function-runtime-provider/lib/utils/legacyBuild.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildResource = void 0;
const execa_1 = __importDefault(require("execa"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const glob_1 = __importDefault(require("glob"));
const path_1 = __importDefault(require("path"));
async function buildResource(request) {
    const resourceDir = path_1.default.join(request.srcRoot, 'src');
    if (!request.lastBuildTimeStamp || isBuildStale(request.srcRoot, request.lastBuildTimeStamp)) {
        installDependencies(resourceDir);
        if (request.legacyBuildHookParams) {
            runBuildScriptHook(request.legacyBuildHookParams.resourceName, request.legacyBuildHookParams.projectRoot);
        }
        return Promise.resolve({ rebuilt: true });
    }
    return Promise.resolve({ rebuilt: false });
}
exports.buildResource = buildResource;
function runBuildScriptHook(resourceName, projectRoot) {
    const scriptName = `amplify:${resourceName}`;
    if (scriptExists(projectRoot, scriptName)) {
        runPackageManager(projectRoot, scriptName);
    }
}
function scriptExists(projectRoot, scriptName) {
    const packageJsonPath = path_1.default.normalize(path_1.default.join(projectRoot, 'package.json'));
    if (fs_extra_1.default.existsSync(packageJsonPath)) {
        const rootPackageJsonContents = require(packageJsonPath);
        return rootPackageJsonContents.scripts && rootPackageJsonContents.scripts[scriptName];
    }
    return false;
}
function installDependencies(resourceDir) {
    runPackageManager(resourceDir);
}
function runPackageManager(cwd, scriptName) {
    const useYarn = fs_extra_1.default.existsSync(`${cwd}/yarn.lock`);
    const packageManager = useYarn ? 'yarn' : 'npm';
    const args = toPackageManagerArgs(useYarn, scriptName);
    try {
        execa_1.default.sync(packageManager, args, {
            cwd,
            stdio: 'pipe',
            encoding: 'utf-8',
        });
    }
    catch (error) {
        if (error.code === 'ENOENT') {
            throw new Error(`Packaging lambda failed function failed. Could not find ${packageManager} executable in the PATH.`);
        }
        else {
            throw new Error(`Packaging lambda failed function failed with the error \n${error.message}`);
        }
    }
}
function toPackageManagerArgs(useYarn, scriptName) {
    if (scriptName) {
        return useYarn ? [scriptName] : ['run-script', scriptName];
    }
    return useYarn ? [] : ['install'];
}
function isBuildStale(resourceDir, lastBuildTimeStamp) {
    const dirTime = new Date(fs_extra_1.default.statSync(resourceDir).mtime);
    if (dirTime > lastBuildTimeStamp) {
        return true;
    }
    const fileUpdatedAfterLastBuild = glob_1.default
        .sync(`${resourceDir}/**`)
        .filter(p => !p.includes('dist'))
        .filter(p => !p.includes('node_modules'))
        .find(file => new Date(fs_extra_1.default.statSync(file).mtime) > lastBuildTimeStamp);
    return !!fileUpdatedAfterLastBuild;
}
