// @aws-amplify/appsync-modelgen-plugin/lib/configs/dart-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DART_RESERVED_KEYWORDS = exports.typeToEnumMap = exports.IGNORE_FOR_FILE = exports.CUSTOM_LINTS_MESSAGE = exports.COLLECTION_PACKAGE = exports.BASE_IMPORT_PACKAGES = exports.LOADER_CLASS_NAME = void 0;
exports.LOADER_CLASS_NAME = 'ModelProvider';
exports.BASE_IMPORT_PACKAGES = [
    'package:flutter/foundation.dart',
    'package:amplify_datastore_plugin_interface/amplify_datastore_plugin_interface.dart',
];
exports.COLLECTION_PACKAGE = 'package:collection/collection.dart';
exports.CUSTOM_LINTS_MESSAGE = `// NOTE: This file is generated and may not follow lint rules defined in your app
// Generated files can be excluded from analysis in analysis_options.yaml
// For more info, see: https://dart.dev/guides/language/analysis-options#excluding-code-from-analysis`;
exports.IGNORE_FOR_FILE = '// ignore_for_file: public_member_api_docs, file_names, unnecessary_new, prefer_if_null_operators, prefer_const_constructors, slash_for_doc_comments, annotate_overrides, non_constant_identifier_names, unnecessary_string_interpolations, prefer_adjacent_string_concatenation, unnecessary_const, dead_code';
exports.typeToEnumMap = {
    String: '.string',
    Int: '.int',
    Float: '.double',
    Boolean: '.bool',
    AWSDate: '.date',
    AWSDateTime: '.dateTime',
    AWSTime: '.time',
    AWSTimestamp: '.timestamp',
    AWSEmail: '.string',
    AWSJSON: '.string',
    AWSURL: '.string',
    AWSPhone: '.string',
    AWSIPAddress: '.string',
};
exports.DART_RESERVED_KEYWORDS = [
    'abstract',
    'as',
    'assert',
    'async',
    'await',
    'break',
    'case',
    'catch',
    'class',
    'const',
    'continue',
    'covariant',
    'default',
    'deferred',
    'do',
    'dynamic',
    'else',
    'enum',
    'export',
    'extends',
    'extension',
    'external',
    'factory',
    'false',
    'final',
    'finally',
    'for',
    'Function',
    'get',
    'hide',
    'if',
    'implements',
    'import',
    'in',
    'interface',
    'is',
    'library',
    'mixin',
    'new',
    'null',
    'on',
    'operator',
    'part',
    'rethrow',
    'return',
    'set',
    'show',
    'static',
    'super',
    'switch',
    'sync',
    'this',
    'throw',
    'true',
    'try',
    'typedef',
    'var',
    'void',
    'while',
    'with',
    'yield',
];
