// @aws-amplify/graphql-transformer-core/lib/transformer-context/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransformerContext = void 0;
const datasource_1 = require("./datasource");
const noop_featuer_flag_1 = require("./noop-featuer-flag");
const output_1 = require("./output");
const provider_registry_1 = require("./provider-registry");
const resolver_1 = require("./resolver");
const resource_helper_1 = require("./resource-helper");
const stack_manager_1 = require("./stack-manager");
class TransformerContext {
    constructor(app, inputDocument, stackMapping, featuerFlags) {
        this.inputDocument = inputDocument;
        this.output = new output_1.TransformerOutput(inputDocument);
        this.resolvers = new resolver_1.ResolverManager();
        this.dataSources = new datasource_1.TransformerDataSourceManager();
        this.providerRegistry = new provider_registry_1.TransformerContextProviderRegistry();
        const stackManager = new stack_manager_1.StackManager(app, stackMapping);
        this.stackManager = stackManager;
        this.resourceHelper = new resource_helper_1.TransformerResourceHelper(stackManager);
        this.featureFlags = featuerFlags !== null && featuerFlags !== void 0 ? featuerFlags : new noop_featuer_flag_1.NoopFeatureFlagProvider();
    }
    bind(api) {
        this._api = api;
        this.resourceHelper.bind(api);
    }
    get api() {
        if (!this._api) {
            throw new Error('API is not initialized till generateResolver step');
        }
        return this._api;
    }
}
exports.TransformerContext = TransformerContext;
